daemon/archive_tarcopyoptions_unix.go

```diff
@@ -14,15 +18,73 @@ func (daemon *Daemon) tarCopyOptions(container *container.Container, noOverwrite
 		return daemon.defaultTarCopyOptions(noOverwriteDirNonDir), nil
 	}
 
-	user, err := idtools.LookupUser(container.Config.User)
+	uid, gid, err := getUIDGID(container.Config.User)
```

Comment:
UID and GID should be resolved from within the container if we want to fix #34142. There's already a function which parses a user[:group] string and resolves it to UID/GID inside the container: [`getUser`](https://github.com/moby/moby/blob/83b51522df43866e39f7c757a6ab84ef4050eeb3/daemon/oci_linux.go#L184-L212). It does delegate to libcontainer/user, though. There's also the [`WithUser`](https://pkg.go.dev/github.com/containerd/containerd/oci#WithUser) function in containerd, though it would be a bit awkward to use here.